### ISSUE TITLE
trying on a test file

### ISSUE THREAD

**Issue by goofballtech** [Anadian/extract-documentation-comments#1]
I was trying to use your library via npx on one of my routes for inline documentation on params and returns. I was getting an error so i mad ea test doc to run against and still l get an error. 

test.js
` 
/** Some test documentation things */
`

command
`npx extract-documentation-comments -n -v -I test.js`

output in console
``` 
npx: installed 65 in 9.722s
note: MainExecutionFunction: Logger: console_stderr transport log level set to: debug
debug: MainExecutionFunction: Soft error: Error: Cannot find module '/home/goofballtech/.npm/_npx/2608/lib/node_modules/extract-documentation-comments/source/package.json'
Require stack:
- /home/goofballtech/.npm/_npx/2608/lib/node_modules/extract-documentation-comments/source/main.js
debug: main_Async: received: [object Object]
info: main_Async: Reading input from a file.
debug: main_Async: options.input: 'test.js'
debug: getDocumentationStringFromSourceString: received: /** Some test documentation things */
,[object Object]
debug: getDocumentationStringFromSourceString: matches: 
debug: getDocumentationStringFromSourceString: returned: 
error: main_Async: "output_string" is either null or not a string.
crit: main_Async: "output_string" is either null or not a string.
debug: MainExecutionFunction: End of execution block.
```

I can only assume that Object to string parse is throwing something off. It looks like the contents of the file are making it into the function fine. I got similar results with the full route file. The contents were displayed and it ended with the same error. I also tried different options in place of -n. Same with each i attempted.

**Comment by goofballtech** [comment]
closed with above PR

**Comment by Anadian** [comment]
Yeah, my original parsing regex assumed that both `/**` and `*/` would be on their own lines so one-liner comments like `/** This won't be extracted. */` wouldn't work; this was simple and had the nice side effect of ensuring that each block of documentation began and ended on its own line but, yeah, this solution was lazy and kind of unintuitive: not to mention I wrote it to assume that `\n` line-feed was what followed the `/**` and it didn't take into account the possibility of `\r` or `\r\n` being used as a line break.

**Comment by Anadian** [comment]
I currently working on a new version as we speak....

**Comment by goofballtech** [comment]
i know how i can implement folder parsing but i think it would be much more efficient if i can break out the main function into sub functions. it would involve some recursion which would be tough in the top to bottom way you have things written now. 

You also seem to have a love affair with try\catch blocks. Goodness man. I have never seen so many so close together. 

I think we can pull some of those out by extracting the path-building and file reading to utility functions where the try\catch is in the function. You can pass your error message object when its called to keep your specific messages. 

thoughts on pulling things out a bit?

**Comment by Anadian** [comment]
Some recursion, specifically to handle a few subdirectories within a source directory, should be fine. It's true, I **love** try/catch blocks; I not particularly partial to the syntax structure itself but unfortunately it is the most universally accepted way to handle errors in Javascript. I find myself using them so frequently that I created a macro in vim to quickly generate them. I'll look into recursively parsing directories myself, I've written code for similar task before, but, as always, you're more than welcome to submit your own solution in a pull request.

**Comment by Anadian** [comment]
I added basic support for multiple input files so, for example, `extract-documentation-comments -I test/*.js -O docs` will extract documentation comments from every JavaScript file in "test/" and write the results for each file to "docs/test/<filename>.js"
This should work for most multi-file cases; of course, I still do plan to add more features. I've pushed this new functionality to GitHub; once I finish making a test for it, I'll release a new npm version with it.

**Comment by goofballtech** [comment]
Not sure how you are doing your local tests before pushing but with the new update i get an error on the folder version.
If i call out a folder as *.js with or without -v all it does is call out the ~~first file~~ **second file** in the folder in the below error
`Unknown value: ./api/routes/baseRoute.js`
command: `npx extract-documentation-comments -I ./api/routes/*.js -O ./api/docs/ -v`

using the above paths i am able to do a single file so the paths are fine.

**Comment by Anadian** [comment]
Can you post the log? Either from STDERR with the -v option or from the log-debug.log file in the log folder. Using the --config option will tell what it's using for the log directory.

**Comment by goofballtech** [comment]
that is the output with -v. 

```
 ✘ ✝  ~/MaintCenter   develop±  npx extract-documentation-comments -I ./api/routes/*.js -O ./api/docs/ -v
Unknown value: ./api/routes/baseRoute.js
 ✘ ✝  ~/MaintCenter   develop±  npx extract-documentation-comments -I ./api/routes/*.js -O ./api/docs/ -v
Unknown value: ./api/routes/baseRoute.js
 ✘ ✝  ~/MaintCenter   develop±  npx extract-documentation-comments -I ./api/routes/*.js -O ./api/docs/ -v --config
Unknown value: ./api/routes/baseRoute.js
 ✘ ✝  ~/MaintCenter   develop±  
```

**Comment by Anadian** [comment]
How are you running the command? From a testing suite like AVA? An automation tool like Grunt or Gulp?
What happens when you try to use it directly from the shell?

**Comment by goofballtech** [comment]
if i go back and run the single file command that works then --config gives me the directories. this is the output in the log when running the failed folder command. 

```
2020-06-28T03:52:53.885Z extract-documentation-comments:ExtractDocumentationComments:extract-documentation-comments.js:MainExecutionFunction:debug: Start of execution block.
2020-06-28T03:52:53.888Z extract-documentation-comments:ExtractDocumentationComments:extract-documentation-comments.js:MainExecutionFunction:note: Logger: console_stderr transport log level set to: debug
2020-06-28T03:52:53.892Z extract-documentation-comments:ExtractDocumentationComments:extract-documentation-comments.js:MainExecutionFunction:debug: Soft error: Error: Cannot find module '/home/goofballtech/extract-documentation-comments/source/package.json'
Require stack:
- /home/goofballtech/extract-documentation-comments/source/main.js
2020-06-28T03:52:53.895Z extract-documentation-comments:ExtractDocumentationComments:extract-documentation-comments.js:MainExecutionFunction:debug: End of execution block.
```

**Comment by goofballtech** [comment]
i am using it directly form the shell via npx in my projects root directory.

**Comment by Anadian** [comment]
Strange, it looks like it's not doing anything; I'll investigate.

**Comment by Anadian** [comment]
Oh, npx is probably using the old version, installed in your node_modules folder, without the multi-file support. `git pull origin master` and try running source/main.js directly through node like: `node ./source/main.js -I [path_to_your_files].js -O docs -v`. It's ugly but it should only be temporary as I'm currently writing a test suite for the new functionality and didn't want to publish a new version to npm before I wrote the test for it.

**Comment by Anadian** [comment]
Any luck?

**Comment by goofballtech** [comment]
I switched over branches in my project from documentation and was working
on adding a new feature. let me move over and fetch your updates to test it
from that repo as there is no easy way to update my project's version until
you push npm.

**Comment by Anadian** [comment]
It's getting pretty late so I'm probably going to have to finish the test suite tomorrow.

**Comment by goofballtech** [comment]
package.json shows version 2.2 as a reference

two separate runs

```
 ✘ ✝  ~/extract-documentation-comments   master±  extract-documentation-comments -I ../MaintCenter/api/routes/*.js -O ../MaintCenter/api/docs
/home/goofballtech/extract-documentation-comments/node_modules/command-line-args/dist/index.js:1345
        throw err
        ^

UNKNOWN_VALUE: Unknown value: ../MaintCenter/api/routes/backupRoute.js
    at commandLineArgs (/home/goofballtech/extract-documentation-comments/node_modules/command-line-args/dist/index.js:1342:21)
    at Object.<anonymous> (/home/goofballtech/extract-documentation-comments/source/main.js:632:16)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  value: '../MaintCenter/api/routes/backupRoute.js'
}
 ✘ ✝  ~/extract-documentation-comments   master±  node ./source/main.js -I ../MaintCenter/api/routes/*.js -O ../MaintCenter/api/docs
/home/goofballtech/extract-documentation-comments/node_modules/command-line-args/dist/index.js:1345
        throw err
        ^

UNKNOWN_VALUE: Unknown value: ../MaintCenter/api/routes/backupRoute.js
    at commandLineArgs (/home/goofballtech/extract-documentation-comments/node_modules/command-line-args/dist/index.js:1342:21)
    at Object.<anonymous> (/home/goofballtech/extract-documentation-comments/source/main.js:632:16)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  value: '../MaintCenter/api/routes/backupRoute.js'
}
```

same output with or without -v flag

**Comment by goofballtech** [comment]
I am headed to bed for the night. If i have time tomorrow i may mess around with moving things into extracted functions to see if i can get it to work without losing your crazy specific error reporting. 

Good Night.

**Comment by Anadian** [comment]
Interestingly, the error is being thrown from within `command-line-args` itself. That's fine, when you get chance, what does the shell print when you try `echo ../MaintCenter/api/routes/*.js`

**Comment by goofballtech** [comment]
```
 ✝  ~/extract-documentation-comments   master±  echo ../MaintCenter/api/routes/*.js
../MaintCenter/api/routes/assetRoute.js ../MaintCenter/api/routes/backupRoute.js ../MaintCenter/api/routes/baseRoute.js ../MaintCenter/api/routes/excelRoute.js ../MaintCenter/api/routes/filesRoute.js ../MaintCenter/api/routes/hoseRoute.js ../MaintCenter/api/routes/index.js ../MaintCenter/api/routes/logRoute.js ../MaintCenter/api/routes/loggerRoute.js ../MaintCenter/api/routes/mailerRoute.js ../MaintCenter/api/routes/meterRoute.js ../MaintCenter/api/routes/settingsRoute.js ../MaintCenter/api/routes/taskRoute.js ../MaintCenter/api/routes/toOrderRoute.js ../MaintCenter/api/routes/todoRoute.js
 ✝  ~/extract-documentation-comments   master±  
```

**Comment by goofballtech** [comment]
why comment out my regex and replace/join to put that block back to the old one?

**Comment by Anadian** [comment]
Your regular expression didn't make much sense: `[\W\w\s\r\n*]` would capture every non-word character, every word character (if you're going to use both, it might as well be a `.` "any" character), every space character (including esoteric ones like `\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff`), carriage return (which is already covered by `\s`), line feed (which is also covered by `\s`), and finally a literal asterisk which would be covered by every `\W` non-word character. There was a lot of redundancy and [you included a lot of characters I don't think you meant to](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes). My version, `[\t\n\r -~]`, is more clear in what it captures: the only control characters it accepts are explicitly tab, line feed, and carriage return, and every "printing" character between and including [ASCII](https://www.asciitable.com/) ` ` space (0x20) and `~` tilde (0x7E).
As for the rest the block, I merely split the long function-chain line to multiple lines to make it more readable.
The only part I got rid of was the `.replace(/\/\*\*|\*\/|(?:\r?\n|\r){2,}/g, '')` as I'm honestly not sure what you're accomplishing with this: so you're replacing `/**`, `*/` as well a series of two or more either `\r?\n` (Unix and Windows-style line breaks) or just `\r` (classic Macintosh-style line break?). First of all, `/**` and `*/` shouldn't be appearing in the match text anyway; this code **was** necessary with your original regex `\/\*\*[\W\w\s\r\n*]*?\*\/` but with mine `/\*\*([\t\n\r -~]*?)\*/`, notice how the actual documentation text being extracted is in a sub match group `(...)`, this way we just take the first sub match and don't have to preform an extra replace operation. As for why you specifically wanted to remove series of two or more line breaks: I have no clue what that accomplishes, is it something specific to how you're writing your documentation?

**Comment by goofballtech** [comment]
With the comment syntax that had /** and */ on multiple line it would pulling comments out with uneven spacing. So i was pulling out multiples an just leaving the on inserted by the join command for consistency. 

I set up 6 or so different syntax examples in regexr.com and tweaked the regex to capture only the comment text no matter what was in it. Based on your own description "It makes zero assumptions about the text between the /** and */" I was setting it up to make zero assumptions. I wanted to grab everything. You are correct i could have removed the group [] and just put a .

If you want to limit it to that at least add the leading and trailing \/ back into the regex as a multi-line comment would break it right now as the typical typical syntax for multi-line comments has a * at the beginning of each line. Any asterisk inside a comment right now ends your regex capture.

**Comment by Anadian** [comment]
Putting an `*` in a documentation comment shouldn't break my regular expression, as `*` is included in the range of printing characters ` -~`, and as far as I can tell, [it doesn't](https://www.dropbox.com/s/f5p40wwd5sl7ma6/Screenshot%20from%202020-06-28%2019-31-29.png?dl=0). Tried a whole bunch of variations on regex101.com, `*` can be placed at the begin, or any where else in the line, just fine.

**Comment by Anadian** [comment]
"add the leading and trailing / back into the regex" 
The forward slashes are already in my regex and have been [this whole time](https://github.com/Anadian/extract-documentation-comments/blob/bb10ec939de3779bf94b5db0aa2b82e916ee51e6/source/main.js): 
```js
	regex = new RegExp('/\\*\\*([\\t\\n\\r -~]*?)\\*/', 'gs');
``` 
Note that the I'm using a string literal to define the regex, not the /../ literal also available, so in the expression above, the `/` at the beginning and end of the string refer to a literal `/` forward slash.

**Comment by Anadian** [comment]
Alright, just published v0.2.3 to npm; it has a shallow but function test suite for multi-file mode. Let me know if you're still having issues.

**Comment by Anadian** [comment]
Hmm, still getting some odd untraceable bugs; I'll have to do some major refactoring tomorrow.

**Comment by Anadian** [comment]
Okay ... I believe I've fixed all the issues in v0.2.4 which I just published to npm.
